### ISSUE TITLE
💡 RFC: Less opinionated, more pluggable MD parsing

### ISSUE THREAD

**Issue by FredKSchott** [snowpackjs/astro#1349]
## Background & Motivation

- We have a few options right now for customizing markdown parsing in Astro:
  - `markdownOptions.remarkPlugins`
  - `markdownOptions.rehypePlugins`
- It's unclear how these options behave, specifically:
  - what is the difference between remark and rehype?
  - Do these completely overwrite Astro's defaults, when I supply my own?
  - What even are the defaults, if I need to add those back?
  - If they get merged, how do they get merged? can I control order?
  - What about stuff like Astro-required markdown processing, does my config conflict with that?
  - what if I want to use something other than remark, rehype?
- It's less flexible to couple our configuration to a single named implementation/tool
  - remark and rehype mean a certain tool, what happens if we switch tools internally?

## Proposal

```
markdown: {
  parse: (content: string) => string; // returns parsed .html/.astro
},
```

1. Add a new `markdown.parse` config options
1. Remove `markdownOptions.remarkPlugins`, `markdownOptions.rehypePlugins`

### Details

- When these config options are not defined, we use our own-builtin parser (no change from current behavior)
- When you define your own, it becomes clear that you are responsible for the entire parsing
- You are free to use whatever parser you want, or even combine in some way
- You are free to share common parser configs, like `markdown.parse = require('astro-markdown-parser-remark')({...})`
- Astro-specific behaviors would be run after the parser, so that this is clearly not the users responsibility.

### Example

By default, we'd have a fully-featured Markdown parser inside of Astro. If you'd like to extend it, you can provide your own parser like this:

```
let parser = unified()
    .use('remark-parse')
    // add your plugins that you want (gfm, slugs, etc)
    // .use('remark-astro') (needed? I don't think so if we implement this as `.astro` output)
    .use('remark-rehype')
    .use('rehype-stringify');

export default ({
  markdown: {
    parse: (content) => parser.process(content).toString()
  }
});
```

### Alternative Designs

```
// easier to extend with more parsers in the future
parsers: {
  markdown: ...
}
```

```
// less risk of organizing in the wrong way, before we know the full config object
markdownParser: ...,
styleParser: ...,
```

### Related RFCs

- Not yet filed, but @natemoo-re and I have talked a bit about moving from a model of "configure everything markdown via the parser" to "configure markdown via custom Astro components". Similar to how MDX lets you define your own H1, H2, Code, etc. components that the parser can then use.

### Open Questions

2. Are there any common use-cases that this feels like overkill for? Ex: adding a single remark plugin? This is a problem with our current API as well, and this proposal offers a solution via `markdown.parse = require('astro-parser-remark')(...)` that could support defaults out of the box.

### Help make it happen!

- [X] I am willing to submit a PR to implement this change.
- [ ] I am willing to submit a PR to implement this change, but would need some guidance.
- [ ] I am not willing to submit a PR to implement this change.

**Comment by drwpow** [comment]
My opinion is as long as Astro’s “no config” story involves _some_ reasonable CSS opinions like scoping out-of-box, I support people fully customizing their CSS and/or overriding it to do as they wish.

I think we have, as always, the 2 apocalyptic horsemen of CSS bundling: Sass’ legacy `@import` behavior and the `postcss-import` PostCSS plugin. Inlining `@imports` is something that people think they want to do, but it just throws a wrench into individual style processing, and that’s something trivial Astro could do for users as an automatic optimization (I’ll skip the example/explanation of why unless it’s helpful).

The former, ideally, Sass users are aware of by now and are aware of how it works or fails. The latter, I think we should just throw a warning if we detect `postcss-import` in a user’s package.json, telling them that Astro handles this? As long as we prevent inlining remote CSS, which is a production build/optimization concern, I think full customization is a great idea.

**Comment by aFuzzyBear** [comment]
Hey guys, I think this addition to the markdown parser and letting users have an element of control over the parsing would be beneficial. 

Is there a away we could let users perhaps write their own plugins aswell? 

Another question that I have, is that does this help pave the way forward for the ellusive `*.mdc`?

**Comment by FredKSchott** [comment]
> My opinion is as long as Astro’s “no config” story involves _some_ reasonable CSS opinions like scoping out-of-box, I support people fully customizing their CSS and/or overriding it to do as they wish.

Agreed, and I think this is worth calling out explicitly: This proposal supports the same or greater level of customization as we have today for both CSS and Markdown. It is more a change to the design of _how_ you customize.

> Is there a away we could let users perhaps write their own plugins aswell?

The nice thing about this is that there's no need for plugins, just use the PostCSS/Unified/Marked/MarkdownIt etc. parser that you'd like directly, following the readme of that project for how to parse/compile input -> output. 

I imagine the community _could_ create plugins to implement this function entirely so that the end user had to do even less, so that you'd have something like:

```
styles: {
  parse: require('astro-postcss-parser')({/* options */})
}
```

> Another question that I have, is that does this help pave the way forward for the ellusive *.mdc?

:D I think @natemoo-re is tackling that this week! Separate from this RFC.

**Comment by aFuzzyBear** [comment]
So very exciting stuff guys.....
@natemoo-re nea pressure huh bud 😅

**Comment by StarfallProjects** [comment]
If parsing is becoming more flexible: it would be awesome to be able to use any markdown parser of choice, or asciidoctor js, as the parser.

**Comment by FredKSchott** [comment]
> If parsing is becoming more flexible: it would be awesome to be able to use any markdown parser of choice, or asciidoctor js, as the parser.

Yes! This is exactly what this RFC hopes to accomplish.

**Comment by matthewp** [comment]
@FredKSchott when discussing offline you mentioned wanting to scope this to markdown for now. Do you want to update the RFC to reflect that?

**Comment by FredKSchott** [comment]
Done! Looking forward to discussing at next RFC call

**Comment by Tc-001** [comment]
Here is an idea:
```js
export default ({
  ".md": {
    parse: (content) => marked(content)
  },
  ".js": {
    parse: (content) => eval(content)
  },
  ".obscure": {
    parse: (content) => custom_parse_and_exec(content)
  }
});
```
It would allow great flexibility in what kind of content can become a page and/or can be imported

Although now that I re-read the convo, it might be a bit out of scope for now :sweat_smile:

**Comment by FredKSchott** [comment]
@Tc-001 you're aligning with something that I think @natemoo-re brought up yesterday, which is: for most assets, this is owned by Vite's build pipeline. Markdown and Styles are the only two interesting ones that can't fit this model since we need to support them as a part of the `.astro` build (`<Markdown>` and `<style>` elements, respectively).

**Comment by matthewp** [comment]
## RFC Call September 28th Notes

- Should probably return an object so metadata/etc can be added.
- Will include remark as the parser by default for now.
- No objections, proceed with this API for markdown.

**Comment by natemoo-re** [comment]
Tracking this as implemented in v0.21.

**Comment by FredKSchott** [comment]
Done in v0.21! Documentation is still needed, post-release.
